### PR TITLE
Add gitea / forgejo support

### PR DESCRIPTION
**Pull request by anbraten**
add gitea support

waiting for:
- https://github.com/go-gitea/gitea/issues/16289
- missing api endpoint for getting a PR by commit

### CODE REVIEW

**Comment by Andre601** [comment]
Curious to know the state of this PR. By the looks of it are the mentioned PRs in gitea merged by now.

File: src/forges/github.ts [hunk 1]
```diff
@@ -36,9 +36,6 @@ export class GithubForge extends Forge {
         repo: options.repo,
         pull_number: pullRequests.data[0].number,
         title: options.title,
-        draft: options.draft,
```
Comment (6543) [inline src/forges/github.ts:39]: unrelated?

**Comment by 6543** [comment]
hmm typecheck is not happy :/

**Comment by anbraten** [comment]
Should be ready to test I think

**Comment by pat-s** [comment]
Can you resolve conflicts and build an image? Happy to test in `woodpecker-plugins`.

**Comment by anbraten** [comment]
Should be available as `anbraten/plugin-ready-release-go:gitea`

File: src/utils/config.ts [hunk 2]
```diff
@@ -13,6 +13,8 @@ const ciConfig = {
   commitMessage: process.env.CI_COMMIT_MESSAGE,
   forgeType: process.env.CI_FORGE_TYPE,
   githubToken: process.env.PLUGIN_GITHUB_TOKEN,
+  giteaToken: process.env.PLUGIN_GITEA_TOKEN,
+  giteaUrl: process.env.PLUGIN_GITEA_URL,
```
Comment (6543) [inline src/utils/config.ts:17]: CI_FORGE_URL should be enouth as default

Comment (anbraten) [reply src/utils/config.ts:17]: Yep, that simplifies the config. (Quite an old PR already 😅 ).

**Comment by pat-s** [comment]
Forgejo support is needed

```
Forge type not supported: forgejo
```

https://ci.codeberg.org/repos/9651/pipeline/311/3

**Comment by anbraten** [comment]
Could you provide a `gitea_token`

**Comment by pat-s** [comment]
https://ci.codeberg.org/repos/9651/pipeline/313/7

Worth bundling `git-lfs`?

**Comment by 6543** [comment]
well this pull conflicts now with target branch :D

**Comment by pat-s** [comment]
We need to test first anyway, doesn't matter for this

**Comment by pat-s** [comment]
@anbraten I see you retriggered the runs and probably also saw the duplicates in https://codeberg.org/woodpecker-plugins/node-pm/pulls/40.

You're polishing everything yourself for the moment? Then I'll go on standby.

**Comment by 6543** [comment]
https://codeberg.org/woodpecker-plugins/node-pm/pulls/40

hmm ... we have duplication and the mentioned pull is a closed non merged one ?!?

File: src/utils/config.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ const ciConfig = {
   commitMessage: process.env.CI_COMMIT_MESSAGE,
   forgeType: process.env.CI_FORGE_TYPE,
   githubToken: process.env.PLUGIN_GITHUB_TOKEN,
+  giteaToken: process.env.PLUGIN_GITEA_TOKEN,
```
Comment (lafriks) [inline src/utils/config.ts:13]: Why not just deprecate GitHub token env and use `token` as there already is forge type so no point in different forge token settings

Comment (6543) [reply src/utils/config.ts:13]: and as woodpecker do tell the plugin already what forge type it is connected to ... we can just deprecate `forgeType` :)

Comment (anbraten) [reply src/utils/config.ts:13]: Will adjust the token name.

It already using woodpeckers forge type: `forgeType: process.env.CI_FORGE_TYPE` 
Nothing we need to deprecate there.

File: src/forges/github.ts (hunk 1 above)
Comment (6543) [reply src/forges/github.ts:39]: moved:  #175 &  #176

**Review by 6543 — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -66,8 +66,10 @@ export default {
 - [x] Support first release (no previous tags)
 - [x] Support defining next version manually (can be done by changing the version on the release pull-request title / commit message)
 - [ ] Support more forges:
-  - [ ] Gitea
+  - [x] Github
+  - [x] Gitea
```
Comment (anbraten) [inline README.md:70]:
```suggestion
  - [x] Gitea / Forgejo
```

**Comment by 6543** [comment]
I approved anbratens part and requested for review for my changes from other maintainers

File: README.md (hunk 4 above)
Comment (Andre601) [reply README.md]: I feel like Forgejo should be treated as its own thing, given it hard-forked and could get breaking changes at any point.

Comment (6543) [reply README.md]: @Andre601 well if they have there own sdk ... 

they currently dont: https://www.npmjs.com/search?q=forgejo

Comment (pat-s) [reply README.md]: @6543 A separate SDK for Forgejo is already in use: https://github.com/woodpecker-ci/woodpecker/blob/b1cf44ae0f80ca18dff3ab67b61cda59107b5a39/go.mod#L12

Comment (pat-s) [reply README.md]: As we already use a separate SDK in WP and the hard-fork has been announced, I also think that treating them individually makes more sense.

Comment (6543) [reply README.md:70]: well I updated the readme ... if there is an acutall forgejo sdk that is not ages old and maintained ... we can switch to it at any time ...

**Comment by 6543** [comment]
> https://codeberg.org/woodpecker-plugins/node-pm/pulls/40 does not yet work correctly. Just reran it. @6543 Why did you approve?

uh sorry i did got your msg worng ... let me see why that one fails :swan:

**Comment by 6543** [comment]
@pat-s works as expected: https://codeberg.org/woodpecker-plugins/test-release-plugin/pulls/3

**Comment by pat-s** [comment]
There are a lot of 404s in this run: https://ci.codeberg.org/repos/9651/pipeline/363/7

File: src/forges/gitea.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  getIssueUrl(owner: string, repo: string, issueNumber: string | number): string {
+    return `${this.url}/${owner}/${repo}/issues/${issueNumber}`;
+  }
+
+  getPullRequestUrl(owner: string, repo: string, pullRequestNumber: string | number): string {
+    return `${this.url}/${owner}/${repo}/pulls/${pullRequestNumber}`;
+  }
+
+  getReleaseUrl(owner: string, repo: string, release: string): string {
+    return `${this.url}/${owner}/${repo}/releases/tags/${release}`;
```
Comment (anbraten) [inline src/forges/gitea.ts:178]: @pat-s Are you sure its `tags`: https://codeberg.org/woodpecker-plugins/node-pm/releases/tag/1.3.0

Comment (pat-s) [reply src/forges/gitea.ts:178]: I thought this references the API endpoint: https://gitea.com/api/swagger#/repository. There it is `tags`. I am currently about to check whether a tag event works with this change (using a custom image). Could be wrong.

Comment (pat-s) [reply src/forges/gitea.ts:178]: OK, my bad, it is not an API path but the final release url. Sorry! In this case it's `tag`. Will build locally and try to narrow down the other issues.

Would really like to get this one finished.

Comment (anbraten) [reply src/forges/gitea.ts]: FYI: https://github.com/go-gitea/gitea/pull/32079

Comment (pat-s) [reply src/forges/gitea.ts]: This might explain all the 404's in https://ci.codeberg.org/repos/13734/pipeline/15/3, but the 

```
# Creating release
Error: undefined
```

when creating a tag is still unresolved.

Even though inferring the PR from a commit fails, the changelog is created and updated as desired.

Comment (anbraten) [reply src/forges/gitea.ts]: @pat-s You did not had releases enabled. Works after enabling it: https://ci.codeberg.org/repos/13734/pipeline/16/3 https://codeberg.org/woodpecker-plugins/test-release-plugin/releases/tag/1.0.0

Comment (pat-s) [reply src/forges/gitea.ts]: 🙈

Comment (pat-s) [reply src/forges/gitea.ts]: Then we're good to go here actually?

Comment (anbraten) [reply src/forges/gitea.ts]: I think so

**Comment by 6543** [comment]
This is still not addressed: https://ci.codeberg.org/repos/13734/pipeline/15/3

**Comment by 6543** [comment]
(It could not create the tag & release)

**Comment by anbraten** [comment]
It is. You forgot to enable releases in codeberg. After enabling it works: https://ci.codeberg.org/repos/13734/pipeline/17/3

**Review by 6543 — APPROVED** [review]
